client/websocket/websocket.go

```diff
@@ -35,65 +35,70 @@ var (
+	return &wsClient{
+		WSLink: ws.NewWSLink(ip, conn, pingPeriod, hndlr),
+		cid:    atomic.AddInt32(&cidCounter, 1),
+	}
+}
+
 // Core specifies the needed methods for Server to operate. Satisfied by *core.Core.
 type Core interface {
 	WalletState(assetID uint32) *core.WalletState
 	SyncBook(dex string, base, quote uint32) (*core.OrderBook, *core.BookFeed, error)
 	AckNotes([]dex.Bytes)
 }
 
-// Server contains fields used by the websocket server.
+// Server is a websocket hub that tracks all running websocket clients, allows
+// sending notifications to all of them, and manages per-client order book
+// subscriptions.
 type Server struct {
-	core       Core
-	log        dex.Logger
-	mSyncerLog dex.Logger
-	ctx        context.Context
-	mtx        sync.RWMutex
+	core Core
+	log  dex.Logger
+	wg   sync.WaitGroup
+
+	clientsMtx sync.RWMutex
 	clients    map[int32]*wsClient
 }
 
-// New returns a new websocket Server. The server ctx is not set upon
-// initiation. It must be set before use by calling Run.
+// New returns a new websocket Server.
 func New(core Core, log dex.Logger) *Server {
 	return &Server{
-		core:       core,
-		log:        log,
-		mSyncerLog: log.SubLogger("MSYNC"),
-		clients:    make(map[int32]*wsClient),
+		core:    core,
+		log:     log,
+		clients: make(map[int32]*wsClient),
 	}
 }
 
-// Run sets the server's context.
-func (s *Server) Run(ctx context.Context) {
-	s.ctx = ctx
-}
-
-type wsClient struct {
-	*ws.WSLink
-	mtx      sync.RWMutex
-	cid      int32
-	feedLoop *dex.StartStopWaiter
-}
-
-func newWSClient(ip string, conn ws.Connection, hndlr func(msg *msgjson.Message) *msgjson.Error) *wsClient {
-	return &wsClient{
-		WSLink: ws.NewWSLink(ip, conn, pingPeriod, hndlr),
-		cid:    atomic.AddInt32(&cidCounter, 1),
-	}
-}
-
-// Shutdown disconnects all connected clients.
+// Shutdown gracefully shuts down all connected clients, waiting for them to
+// disconnect and any running goroutines and message handlers to return.
 func (s *Server) Shutdown() {
-	s.mtx.Lock()
+	s.clientsMtx.Lock()
 	for _, cl := range s.clients {
 		cl.Disconnect()
 	}
-	s.mtx.Unlock()
+	s.clientsMtx.Unlock()
+	// Each upgraded connection handler must return. This waits for running
+	// marketSyncers, and response handlers such as long as
+	// dex/ws.(*WSLink).{Connect,inHandler} operate as designed.
+	s.wg.Wait()
 }
 
-// HandleConnect handles the websocket connection request, creating a ws.Connection
-// and a connect thread.
-func (s *Server) HandleConnect(w http.ResponseWriter, r *http.Request) {
+// HandleConnect handles the websocket connection request, creating a
+// ws.Connection and a connect thread. Since the http.Request's Context is
+// canceled after ServerHTTP returns, a separate context must be provided to be
+// able to cancel the hijacked connection handler at a later time since this
+// function is not blocking.
+func (s *Server) HandleConnect(ctx context.Context, w http.ResponseWriter, r *http.Request) {
```

Comment:
There's no reason each request's context has to be the same, and this makes the context more request scoped, even if the caller (webserver/rpcserver) uses their main Connect contexts for each upgrade currently.